### PR TITLE
Make wav output duration adaptive to input length (word count)

### PR DESCRIPTION
**Pull request by mertyildiran**

### CODE REVIEW

**Comment by keithito** [comment]
Thanks for sending this PR, and apologies for taking so long to get back to you!

This is a very interesting approach, but I'm not sure if rebuilding the the model on every input is necessarily the right solution to this problem. Two approaches that I suspect might work are:

1. As a simple solution, trimming silence from the output using [librosa.effects.trim](https://librosa.github.io/librosa/generated/librosa.effects.trim.html) or something similar.

2. Adding a classifier to the model that predicts based on the decoder RNN state when to stop decoding.

That said, I could be wrong, and if there's a convincing argument that this is the best approach, I'd be happy to accept this PR. Please let me know!

**Comment by mertyildiran** [comment]
@keithito trimming is not a good solution as far as I understand from the tests I have made. Not fixed `hparams.max_iters` value is affecting the output badly, like stretching some vowels etc... Making `hparams.max_iters` adaptive is increasing the speed, in other words decreasing the response time. Also what if user tries to give a long input worth of more than 6 seconds, like a long paragraph? I tried that, with `hparams.max_iters = 100` it's giving me only first 6 seconds of the audio. So `hparams.max_iters` must be adaptive.

The best argument against this PR would be: "Why did you use 130 words per minute constant because it's specific to English?" Because it will cause some problems when you (for example) try to use tacotron for Chinese. I don't know any language independent approach to this problem, so after this PR we should probably add language specific exceptions.

**Comment by keithito** [comment]
Can you try patching commit faf7b8d0 and seeing if it works for you? It increases max_iters to 200 and trims the audio. This means the output can be up to about 13 sec with the default hyperparameters, and there's not a huge increase in synthesis time if you're running on a GPU. While this still imposes a limit, I don't think models will scale to outputs that are much longer than what they've seen in the training data anyway.

**Comment by mertyildiran** [comment]
@keithito here are the comparisons of https://github.com/keithito/tacotron/commit/faf7b8d083c163f77a5be793350f36558695600b and this pull request:

> **Text:** Once upon a time there was a dear little girl who was loved by every one who looked at her, but most of all by her grandmother, and there was nothing that she would not have given to the child. Once she gave her a little cap of red velvet, which suited her so well that she would never wear anything else. So she was always called Little Red Riding Hood. - [Little Red Riding Hood](https://americanliterature.com/childrens-stories/little-red-riding-hood)

 - [Long Speech generated by tacotron trim-audio branch](https://clyp.it/shrs2en3)
 - [Long Speech generated by tacotron adaptive PR](https://clyp.it/02qijkdr)

> **Text:** hello world

 - [Hello World generated by tacotron trim-audio branch](https://clyp.it/fhh4ffx4)
 - [Hello World generated by tacotron adaptive PR](https://clyp.it/mlz53tvg)

[My profile link](https://clyp.it/user/1d01ehwr) to see all of them side by side.

First of all, tacotron is currently not doing well against long inputs but it's not related to this pull request.

For short inputs this PR is fine. But for long inputs, I think, we should combine this PR with the trimming out the silent part at the end (the solution you have said). After this PR, we can add the code about trimming out so the most of the issues related with duration will be solved.

*P.S. I used the [official model](http://data.keithito.com/data/speech/tacotron-20170720.tar.bz2) (of yours) to generate these audio outputs.*

**Comment by keithito** [comment]
Thanks for doing the comparisons! I updated the audio trimming to trim at the first long silence. This should fix the issue when there's garbage later on in the output.

I still don't think that we should be updating the model and reloading the checkpoint for every input. I tried it out, and it's a bit slow, and the word count heuristic sometimes leads to cutting off the output (for example in the 5th eval sentence, "Generative adversarial network or variational auto-encoder."). Also, though the Synthesizer doesn't make use of it, the model supports synthesizing a batch of inputs at once, and I'm not sure how that would fit in with this change.

**Comment by mertyildiran** [comment]
@keithito first of all, because [Tacotron](https://arxiv.org/abs/1703.10135) is an **End-to-End** solution, you cannot limit the sequence length with a constant `hparams.max_iters`. Also can't you see [this hello world](https://clyp.it/mlz53tvg) sounds better than [that hello world](https://clyp.it/fhh4ffx4)?

For your "it's slow" claim, here are the response time comparisons:

> **Text:** hello world

Commit or PR | Response Time | Output
--- | --- | ---
https://github.com/keithito/tacotron/commit/faf7b8d083c163f77a5be793350f36558695600b | 0:01:19.447077 | 12s
https://github.com/keithito/tacotron/commit/9fc433a8702a5e8392254f4c2cc6fe035c64e424 | 0:01:21.199202 | 1s
this pull request | 0:00:28.621482 | 1s

> **Text:** Once upon a time there was a dear little girl who was loved by every one who looked at her, but most of all by her grandmother, and there was nothing that she would not have given to the child. Once she gave her a little cap of red velvet, which suited her so well that she would never wear anything else. So she was always called Little Red Riding Hood. - [Little Red Riding Hood](https://americanliterature.com/childrens-stories/little-red-riding-hood)

Commit or PR | Response Time | Output
--- | --- | ---
https://github.com/keithito/tacotron/commit/faf7b8d083c163f77a5be793350f36558695600b | 0:01:16.363043 | 12s
https://github.com/keithito/tacotron/commit/9fc433a8702a5e8392254f4c2cc6fe035c64e424 | 0:01:21.524000 | 12s
this pull request | 0:03:40.034392 | 34s

But this is not a good comparison. while your commits are limiting the output to 12 seconds, this pull request predicted 34 seconds worth of reading, Google TTS read it in 22 seconds and I read it in 29 seconds. So in reality there is approximately 28 seconds worth of reading for this *Little Red Riding Hood* text. (*Google TTS is intentionally designed to be a bit quicker than normal reading speed...*) 

> **Text:** Generative adversarial network or variational auto-encoder.

Commit or PR | Response Time | Output
--- | --- | ---
https://github.com/keithito/tacotron/commit/faf7b8d083c163f77a5be793350f36558695600b | 0:01:16.014426 | 4s
https://github.com/keithito/tacotron/commit/9fc433a8702a5e8392254f4c2cc6fe035c64e424 | 0:01:19.364656 | 4s
this pull request | 0:00:49.981970 | 4s

That's a cherry picked example that mostly consists of **phonetically long words**. So for this example, I added a buffer that worth of two seconds to handle these kind of situations (to my local clone):

`hparams.max_iters = round(1000 * (float(len(req.params.get('text').split())) / 130)) + 33`

**+ 33** is the two seconds buffer here. But you should be the one who determine the length of the buffer because you are the most experienced person on the subject. Because of this reason I did not include a buffer to this pull request.

In conclusion, this pull request is clearly faster and more tolerant when compared to the other solutions. Also see how the response time of this pull request is linearly correlated to the input length. With the stack of this pull request and https://github.com/keithito/tacotron/commit/d88640b7c0003923df3ee3b79155205997ec0bd7 the most of the issues related with duration will be solved.

**Comment by keithito** [comment]
@mertyildiran, Tacotron is end-to-end in the sense that it's a single neural network that goes from (almost) raw text to (almost) waveform output. It's not fully end-to-end since you currently have to do some text processing at the start and Griffin-Lim to convert from a spectrogram to wav at the end. I don't see why this would preclude making practical choices like enforcing a maximum number of decoder steps.

At least for now, when synthesizing passages of text that are much longer than what was seen in the training data, you're going to want to chop them up at sentence boundaries. Whether there's a hard limit on output length or not, you're not going to get good results with this model on long paragraphs.

Regarding quality, the "Hello World" generated by [master](https://clyp.it/04mlx4ip) and [this PR](https://clyp.it/j15kpsdr) sound almost identical to me. I can't really say one is better than the other.

Regarding speed, here are my timings on a GTX 1080 Ti:

> **Text:** hello world
Branch | Response Time
--- | ---
master | 0.598 sec
PR #43 | 14.75 sec

> **Text:** Generative adversarial network or variational auto-encoder.
Branch | Response Time
--- | ---
master | 0.721 sec
PR #43 | 18.14 sec

> **Text:** Once upon a time there was a dear little girl who was loved by every one who looked at her, but most of all by her grandmother, and there was nothing that she would not have given to the child.

*note: I only included the first sentence (~10 sec) of this one so the timing is a fair comparison*

Branch | Response Time
--- | ---
master | 0.587 sec
PR #43 | 21.44 sec

**Comment by keithito** [comment]
For what it's worth, I'm not necessarily opposed to allowing an option where max_iters is dynamically set, but perhaps it would be possible to pass the value down as a tensor so you don't need to rebuild the model in order to change it?

**Comment by pineking** [comment]
@keithito we rewrite `the tf.contrib.seq2seq.dynamic_decode` function, so the max_iters value can be changed dynamically, and don't need to rebuild/load the model.

**Comment by mertyildiran** [comment]
@keithito OK, mine were CPU results and when it comes to GPU, rebuilding the model is becoming the bottleneck. I will try to update this pull request. Please give me some time...

**Comment by mertyildiran** [comment]
@keithito I realized that the primary bottleneck was an unnecessary reload of the checkpoint (my bad) and I removed it with my second commit. It turns out that updating the model is not even computationally expensive, so that was not the problem.

Secondary factor that causing the speed problem was [excessive amount of iterations](https://github.com/keithito/tacotron/blob/master/util/audio.py#L87-L90) is being done inside the implementation of Griffin-Lim algorithm. So I reduced the constant from 60 to 10.

Now the speed and audio quality are indistinguishable from the original state of the repository, and the output duration is adaptive to input length. Please run your tests one last time to validate the results. :blush:

**Comment by keithito** [comment]
That's great to hear! :-) I'll take a look tonight.

**Comment by keithito** [comment]
Hi @mertyildiran, this is definitely faster than before, but still seems slower than master:

> **Text:** hello world

Configuration | Synthesis Time
--- | ---
master (max_iters=200, griffin_lim_iters=60) | 0.605 sec
master (max_iters=500, griffin_lim_iters=60) | 1.287 sec
This PR (griffin_lim_iters=10)               | 2.665 sec
This PR (griffin_lim_iters=60)               | 10.976 sec


> **Text:** Generative adversarial network or variational auto-encoder.

Configuration | Synthesis Time
--- | ---
master (max_iters=200, griffin_lim_iters=60) | 0.731 sec
master (max_iters=500, griffin_lim_iters=60) | 1.375 sec
This PR (griffin_lim_iters=10)               | 3.081 sec
This PR (griffin_lim_iters=60)               | 12.668 sec


> **Text:** Once upon a time there was a dear little girl who was loved by every one who looked at her, but most of all by her grandmother, and there was nothing that she would not have given to the child.

Configuration | Synthesis Time
--- | ---
master (max_iters=200, griffin_lim_iters=60) | 0.669 sec
master (max_iters=500, griffin_lim_iters=60) | 1.466 sec
This PR (griffin_lim_iters=10)               | 3.912 sec
This PR (griffin_lim_iters=60)               | 14.204 sec


Perhaps 60 Griffin-Lim iterations is overkill (they use 50 in the Tacotron paper), but 10 is too few. You can hear artifacts in the samples with 10 iterations.

However, the speed is only one issue. The more significant issue, at least in my mind, is that this change adds some assumptions and complexity to the model:

1. It introduces assuptions about how fast people speak and how long a word is. There's a large difference in speaking speed between datasets (e.g. LJ Speech vs Nancy). There's even wide variation between individual speakers in the VCTK dataset. There are also differences in word length and how words are separated between languages. Trying to guess this with a heuristic seems a bit awkward.

2. It modifies the model for every eval case. I don't think it's how Tensorflow is intended to be used, and I don't know of any other Tensorflow models that do this. At some point, I would like to use Tensorflow's meta graph support to load the graph from the checkpoint, which will make training and loading multiple models with custom hyperparameters easier, but I think doing so would be incompatible with this change.

3. It makes the code more complex. Now there's a model building step and a model updating step, and some of the code is copied between the two and some of it is not. Now when making a change, you might have to make a change in both places, etc.


In some cases, it's worth it to add complexity and additional assumptions if they provide  compelling improvements. So I wanted to spend some time experimenting and trying to understand the problems that this PR solves (I didn't have time to do so until just now). Unfortunately, I'm having a hard time seeing what these improvements are.

You can already set max_iters to a value that's much larger than the model can handle (400 or 500 iters?) with a relatively minor performance penalty when running on a GPU. As far as I can tell, doing so does not negatively impact quality on shorter phrases. All of the phrases I tried sound the same with this PR and master, except for ones longer than 30 seconds where they both generate gibberish, but master's gibberish is shorter.

**Comment by mertyildiran** [comment]
@keithito my results with NVIDIA GeForce GT 640 (not everyone has world's most expensive GPU):

> Text: hello world

Configuration | Synthesis Time
-- | --
master (max_iters=200, griffin_lim_iters=60) | 1.826 sec
master (max_iters=500, griffin_lim_iters=60) | 4.628 sec
This PR (griffin_lim_iters=10) | 4.716 sec
This PR (griffin_lim_iters=60) | 19.244 sec

> Text: Generative adversarial network or variational auto-encoder.

Configuration | Synthesis Time
-- | --
master (max_iters=200, griffin_lim_iters=60) | 2.397 sec
master (max_iters=500, griffin_lim_iters=60) | 4.694 sec
This PR (griffin_lim_iters=10) | 5.519 sec
This PR (griffin_lim_iters=60) | 20.274 sec

> Text: Once upon a time there was a dear little girl who was loved by every one who looked at her, but most of all by her grandmother, and there was nothing that she would not have given to the child.

Configuration | Synthesis Time
-- | --
master (max_iters=200, griffin_lim_iters=60) | 2.319 sec
master (max_iters=500, griffin_lim_iters=60) | 4.784 sec
This PR (griffin_lim_iters=10) | 7.718 sec
This PR (griffin_lim_iters=60) | 23.394 sec

I don't know how you are planning to deal with long inputs (both in training and evaluating) but I gave up on contributing to this project. [Festival](http://www.cstr.ed.ac.uk/projects/festival/)(with cmu_us_clb_arctic voice) is already synthesising very natural sounding speech without any GPU need so I will continue to use Festival. https://youtu.be/krHUzY2DylI
